Run main rust-analyzer tests in rust-lang/rust CI

jieyouxu opened the pull request:
Part of rust-lang/rust#147370.
MCP: https://github.com/rust-lang/compiler-team/issues/923

This PR prepares `rust-analyzer` crates with `in-rust-tree` cargo featues where needed, and and updates bootstrap to run the main `rust-analyzer` tests in rust-lang/rust CI, not just the `proc-macro-srv` crate tests.

This supersedes the earlier attempt at https://github.com/rust-lang/rust/pull/136779. I was honestly expecting more failures in this PR, but looking back at the previous attempt, that makes sense because we no longer run `i686-mingw` (32-bit windows-gnu) which had a _bunch_ of these failures. In the earlier attempt I also disabled the `i686-mingw`-related failures for `i686-msvc` since I didn't feel like digging into 32-bit msvc at the time. Try results from this PR shows that it's most likely limited to 32-bit windows-gnu specifically.

### `rust-analyzer` test remarks

- I actually had to _remove_ the `CARGO_WORKSPACE_DIR` `expect-test`-hack in order for `expect-test` to be able to find the test expectation HTML files (for `syntax_highlighting` tests in `ide`). When I added the hack, ironically, it made `expect-test` unable to find the expectation files. I think this was because previously the path was of the `proc-macro-srv` crate specifically, now we point to the root r-a workspace?
- The `cfg`-related differences on `aarch64-apple-darwin` might've been fixed? I can't tell, but we don't seem to be observing the differences now.
- I'm not sure why `config::{generate_config_documentation, generate_package_json_config}` no longer fails. Perhaps they were fixed to no longer try to write to source directory?

### Review remarks

- Commit 1 updates r-a crates that are involved in tests needing artifacts from `rustc_private` compiler crates to use the `in-rust-tree` cargo feature. I briefly tried to use a plain `--cfg=in_rust_tree`, but quickly realized it was very hacky, and needed invasive bootstrap changes. The cargo feature approach seems most "natural"/well-supported to both bootstrap and cargo.
- Commit 2 updates bootstrap to not only run the `proc-macro-srv` tests, but the whole r-a tests.
- Commit 3 restricts r-a main tests to non-32-bit targets we test in CI, since (1) r-a repo does not run tests against 32-bit platforms, and (2) there are some target pointer width sensitive hash differences causing tests to fail. Notably, this means that we also no longer run r-a `proc-macro-srv` tests against 32-bit targets, but we don't expect that crate to be have target pointer width differences. Discussed this in [#t-compiler/rust-analyzer > 32-bit tests?](https://rust-lang.zulipchat.com/#narrow/channel/185405-t-compiler.2Frust-analyzer/topic/32-bit.20tests.3F/with/563145736).

---

// try-job: aarch64-gnu
// try-job: aarch64-apple
// try-job: x86_64-mingw-1
// try-job: i686-msvc-1
// try-job: x86_64-msvc-1
// try-job: aarch64-msvc-1

jieyouxu committed:
bootstrap: run full `rust-analyzer` tests

jieyouxu committed:
rust-analyzer: prep crates for testing against in-tree `rustc_private`

jieyouxu commented:
@bors try

jieyouxu commented:
I'm a bit suspicious given prior experience in https://github.com/rust-lang/rust/pull/136779

jieyouxu commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -446,35 +446,61 @@ impl Step for RustAnalyzer {
+        // FIXME: RA's test suite tries to write to the source directory, that can't work in Rust CI
+        // without properly wiring up the writable test dir.
         cargo.env("SKIP_SLOW_TESTS", "1");
+        // FIXME: teach rust-analyzer to use `RUST_ANALYZER_TEST_DIR` for its test output root
+        // directory. In the rust-lang/rust CI, we separate checkout directory vs build directory,
+        // where the checkout directory is read-only whereas build directory (including test output
+        // directories) is writable.
+        let dir = testdir(builder, target);
+        t!(fs::create_dir_all(&dir));
+        cargo.env("RUST_ANALYZER_TEST_DIR", dir);
```
TODO: I added this mostly for review discussion, remove before merge if r-a wants to use another approach or don't need this.

jieyouxu commented:
r? @Kobzol (for bootstrap side)
r? @Veykril (for r-a side)

FYI @rust-lang/rust-analyzer

Kobzol commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -446,35 +446,54 @@ impl Step for RustAnalyzer {
+        // NOTE: we need to skip `src/tools/rust-analyzer/xtask` as they seem to exercise rustup /
+        // stable rustfmt.
+        //
+        // NOTE: you can only skip a specific workspace package via `--skip=...` if you *also*
```
The comment mentions `--skip`, but the code uses `--exclude`, that's a bit confusing.

jieyouxu replied:
`--exclude` is what the comment should've said, fixed in https://github.com/rust-lang/rust/compare/bbfdd244ef9beae0be09fa559e5a05fb2a4eb12c..bb43acaef72057aec5e778724e6b0c764e189bfe.

Kobzol commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -446,35 +446,54 @@ impl Step for RustAnalyzer {
+        let build_compiler = tool_result.build_compiler;
+        let target = self.compilers.target();
+
+        let test_compiler = self.compilers.target();
```
```suggestion
        let target = self.compilers.target();
```
The variable is duplicated now.

Kobzol reviewed (commented):
The changes look fine, but I'm not sure if this doesn't require an MCP. We should also check the CI durations.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147372/1a3cdd34629306fa67624eaa60d73687e7fcf855..bbfdd244ef9beae0be09fa559e5a05fb2a4eb12c)*

jieyouxu commented:
> The changes look fine, but I'm not sure if this doesn't require an MCP.

I will open one if just for the visibility, even if in the end we find that it's not technically needed.

jieyouxu commented on src/tools/rust-analyzer/crates/ide-db/Cargo.toml:
```diff
@@ -53,5 +53,13 @@ line-index.workspace = true
 
+# local deps
+test-utils.workspace = true
+test-fixture.workspace = true
```
Fixed.

bors commented:
:umbrella: The latest upstream changes (presumably #149265) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Veykril commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -445,35 +445,61 @@ impl Step for RustAnalyzer {
+            "tests::smoke_test_real_sysroot_cargo",
+            // NOTE: part of `smol-str` test suite; this tries to access a stable rustfmt from the
+            // environment, which is not something we want to do.
+            "check_code_formatting",
```
Note: It would be fine to skip running the tests in the `lib` dir I think. Those crates won't ever depend on anything in-tree

jieyouxu commented:
> Note: It would be fine to skip running the tests in the `lib` dir I think. Those crates won't ever depend on anything in-tree

Pushed a change to [exclude crates under `lib/` instead](https://triagebot.infra.rust-lang.org/gh-range-diff/rust-lang/rust/6840234806e4a57e14112a1137783c885b7a23db..f1944b6aa975e3ec9aa1f67d2b4762dcc4a85690/1be6b13be73dc12e98e51b403add4c41a0b77759..86bcd31bd8c9dd8bd2374abdf0c8fb00d8fb6f44).

@rustbot ready

Kobzol commented:
Oh wow, the diff is like, way smaller than I thought :facepalm: Sorry for the large delay, I had too much IRL stuff in the recent weeks to focus on reviews.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19914050449/job/57089026672) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/57089026672) [(plain)](https://github.com/rust-lang/rust/commit/990871031de95518298827e25210dea8fd6f8e53/checks/57089026672/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: all_stage_1
Source: src/bootstrap/src/core/builder/tests.rs:2078
────────────────────────────────────────────────────────────────────────────────
Expression: prepare_test_config(&ctx).render_steps()
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
   54    54 │ [doc] rustc 0 <host> -> releases 1 <host>
   55    55 │ [build] rustc 0 <host> -> Linkchecker 1 <host>
   56    56 │ [test] link-check <host>
   57    57 │ [test] tier-check <host>
   58       │-[build] rustc 0 <host> -> rust-analyzer 1 <host>
   59    58 │ [test] rustc 0 <host> -> rust-analyzer 1 <host>
   60    59 │ [build] rustc 0 <host> -> RustdocTheme 1 <host>
   61    60 │ [test] rustdoc-theme 1 <host>
   62    61 │ [test] compiletest-rustdoc-ui 1 <host>
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.

thread 'core::builder::tests::snapshot::test_all_stage_1' (13381) panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:679:13:
snapshot assertion for 'all_stage_1' failed in line 2078
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/std/src/panicking.rs:698:5
   1: core::panicking::panic_fmt
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/panicking.rs:80:14
   2: <insta::runtime::SnapshotAssertionContext>::finalize
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:679:13
   3: insta::runtime::assert_snapshot
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:848:13
   4: bootstrap::core::builder::tests::snapshot::test_all_stage_1
             at ./src/core/builder/tests.rs:2078:9
   5: bootstrap::core::builder::tests::snapshot::test_all_stage_1::{closure#0}
             at ./src/core/builder/tests.rs:2076:26
   6: <bootstrap::core::builder::tests::snapshot::test_all_stage_1::{closure#0} as core::ops::function::FnOnce<()>>::call_once
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/ops/function.rs:250:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---- core::builder::tests::snapshot::test_all_stage_1 stdout end ----
---- core::builder::tests::snapshot::test_all_stage_2 stdout ----
Running test in /tmp/.tmpfG8WAc
##[group]Building stage1 tidy (stage0 -> stage1, aarch64-unknown-linux-gnu)
##[endgroup]
fmt check
tidy check
x.py completions check
---
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: all_stage_2
Source: src/bootstrap/src/core/builder/tests.rs:2259
────────────────────────────────────────────────────────────────────────────────
Expression: prepare_test_config(&ctx).stage(2).render_steps()
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
   57    57 │ [doc] rustc 1 <host> -> releases 2 <host>
   58    58 │ [build] rustc 0 <host> -> Linkchecker 1 <host>
   59    59 │ [test] link-check <host>
   60    60 │ [test] tier-check <host>
   61       │-[build] rustc 1 <host> -> rust-analyzer 2 <host>
   62    61 │ [test] rustc 1 <host> -> rust-analyzer 2 <host>
   63    62 │ [doc] rustc (book) <host>
   64    63 │ [test] rustc 1 <host> -> lint-docs 2 <host>
   65    64 │ [build] rustc 0 <host> -> RustdocTheme 1 <host>
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.

thread 'core::builder::tests::snapshot::test_all_stage_2' (13382) panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:679:13:
snapshot assertion for 'all_stage_2' failed in line 2259
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/std/src/panicking.rs:698:5
   1: core::panicking::panic_fmt
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/panicking.rs:80:14
   2: <insta::runtime::SnapshotAssertionContext>::finalize
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:679:13
   3: insta::runtime::assert_snapshot
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:848:13
   4: bootstrap::core::builder::tests::snapshot::test_all_stage_2
             at ./src/core/builder/tests.rs:2259:9
   5: bootstrap::core::builder::tests::snapshot::test_all_stage_2::{closure#0}
             at ./src/core/builder/tests.rs:2257:26
   6: <bootstrap::core::builder::tests::snapshot::test_all_stage_2::{closure#0} as core::ops::function::FnOnce<()>>::call_once
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/ops/function.rs:250:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---- core::builder::tests::snapshot::test_all_stage_2 stdout end ----

failures:
    core::builder::tests::snapshot::test_all_stage_1
    core::builder::tests::snapshot::test_all_stage_2

```

</details>

Kobzol commented:
Thanks, let's try. A sync has happened in the meantime, I guess? :laughing: You can r=me once you rebase.

@bors rollup=never

jieyouxu commented:
Oh no worries, thankfully these merge conflicts are easy to fix 😁

jieyouxu commented:
(Fantastic, I pushed an outdated branch xD)

rust-log-analyzer commented:
The job **`i686-gnu-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19958935918/job/57234459086) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/57234459086) [(plain)](https://github.com/rust-lang/rust/commit/cd72010918e4428571b73631d12653b8bf8cabb5/checks/57234459086/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   --> src/tools/rust-analyzer/crates/hir-def/src/item_tree/tests.rs:73:9

You can update all `expect!` tests by running:

    env UPDATE_EXPECT=1 cargo test

To update a single test, place the cursor on `expect` token and use `run` feature of rust-analyzer.

Expect:
----
#[on_extern_block]
// AstId: ExternBlock[0000, 0]
extern {
    #[on_extern_type]
    // AstId: TypeAlias[A09C, 0]
    pub(self) type ExType;

    #[on_extern_static]
    // AstId: Static[D85E, 0]
    pub(self) static EX_STATIC = _;

    #[on_extern_fn]
    // AstId: Fn[B240, 0]
    pub(self) fn ex_fn;
}

----

Actual:
----
#[on_extern_block]
// AstId: ExternBlock[0000, 0]
extern {
    #[on_extern_type]
    // AstId: TypeAlias[E072, 0]
    pub(self) type ExType;

    #[on_extern_static]
    // AstId: Static[6AFE, 0]
    pub(self) static EX_STATIC = _;

    #[on_extern_fn]
    // AstId: Fn[5255, 0]
    pub(self) fn ex_fn;
}

----

Diff:
----
#[on_extern_block]
// AstId: ExternBlock[0000, 0]
extern {
    #[on_extern_type]
    // AstId: TypeAlias[A09CE072, 0]
    pub(self) type ExType;

    #[on_extern_static]
    // AstId: Static[D856AFE, 0]
    pub(self) static EX_STATIC = _;

    #[on_extern_fn]
    // AstId: Fn[B2405255, 0]
    pub(self) fn ex_fn;
}

----

---- item_tree::tests::extern_blocks stdout end ----
---
   --> src/tools/rust-analyzer/crates/hir-def/src/item_tree/tests.rs:123:9

Expect:
----
// AstId: Struct[ED35, 0]
pub(self) struct Unit;

#[derive(Debug)]
// AstId: Struct[A47C, 0]
pub(self) struct Struct { ... }

// AstId: Struct[C8C9, 0]
pub(self) struct Tuple(...);

// AstId: Union[2797, 0]
pub(self) union Ize { ... }

// AstId: Enum[7D23, 0]
pub(self) enum E { ... }

----

Actual:
----
// AstId: Struct[5ECD, 0]
pub(self) struct Unit;

#[derive(Debug)]
// AstId: Struct[E715, 0]
pub(self) struct Struct { ... }

// AstId: Struct[FC08, 0]
pub(self) struct Tuple(...);

// AstId: Union[3086, 0]
pub(self) union Ize { ... }

// AstId: Enum[2360, 0]
pub(self) enum E { ... }

----

Diff:
----
// AstId: Struct[ED355ECD, 0]
pub(self) struct Unit;

#[derive(Debug)]
// AstId: Struct[A47CE715, 0]
pub(self) struct Struct { ... }

// AstId: Struct[C8C9FC08, 0]
pub(self) struct Tuple(...);

// AstId: Union[27973086, 0]
pub(self) union Ize { ... }

// AstId: Enum[7D2360, 0]
pub(self) enum E { ... }

----

---- item_tree::tests::adts stdout end ----
---- item_tree::tests::imports stdout ----
---
Expect:
----
#![no_std]

// AstId: ExternCrate[070B, 0]
pub(self) extern crate self as renamed;

// AstId: ExternCrate[1EA5, 0]
pub(in super) extern crate bli;

// AstId: Use[0000, 0]
pub use crate::path::{nested, items as renamed, Trait as _};

// AstId: Use[0000, 1]
pub(self) use globs::*;

// AstId: Use[0000, 2]
pub(self) use crate::{A, B};

// AstId: Use[0000, 3]
pub(self) use a::{c, d::{e}};

----

Actual:
----
#![no_std]

// AstId: ExternCrate[E25E, 0]
pub(self) extern crate self as renamed;

// AstId: ExternCrate[EEF2, 0]
pub(in super) extern crate bli;

// AstId: Use[0000, 0]
pub use crate::path::{nested, items as renamed, Trait as _};

// AstId: Use[0000, 1]
pub(self) use globs::*;

// AstId: Use[0000, 2]
pub(self) use crate::{A, B};

// AstId: Use[0000, 3]
pub(self) use a::{c, d::{e}};

----

Diff:
----
#![no_std]

// AstId: ExternCrate[070BE25E, 0]
pub(self) extern crate self as renamed;

// AstId: ExternCrate[1EA5EEF2, 0]
pub(in super) extern crate bli;

// AstId: Use[0000, 0]
pub use crate::path::{nested, items as renamed, Trait as _};

// AstId: Use[0000, 1]
pub(self) use globs::*;

// AstId: Use[0000, 2]
pub(self) use crate::{A, B};

// AstId: Use[0000, 3]
pub(self) use a::{c, d::{e}};

----

---- item_tree::tests::imports stdout end ----
---- item_tree::tests::macros stdout ----


error: expect test failed
   --> src/tools/rust-analyzer/crates/hir-def/src/item_tree/tests.rs:222:9

Expect:
----
// AstId: MacroRules[7E68, 0]
macro_rules! m { ... }

// AstId: MacroDef[1C1E, 0]
pub macro m2 { ... }

// AstId: MacroCall[7E68, 0], SyntaxContextId: ROOT2024, ExpandTo: Items
m!(...);

----

Actual:
----
// AstId: MacroRules[7D0B, 0]
macro_rules! m { ... }

// AstId: MacroDef[A602, 0]
pub macro m2 { ... }

// AstId: MacroCall[7D0B, 0], SyntaxContextId: ROOT2024, ExpandTo: Items
m!(...);

----

Diff:
----
// AstId: MacroRules[7E68D0B, 0]
macro_rules! m { ... }

// AstId: MacroDef[1C1EA602, 0]
pub macro m2 { ... }

// AstId: MacroCall[7E68D0B, 0], SyntaxContextId: ROOT2024, ExpandTo: Items
m!(...);

----

---- item_tree::tests::macros stdout end ----
---- item_tree::tests::modules stdout ----


error: expect test failed
   --> src/tools/rust-analyzer/crates/hir-def/src/item_tree/tests.rs:194:9

Expect:
----
// AstId: Module[03AE, 0]
pub(self) mod inline {
    // AstId: Use[0000, 0]
    pub(self) use super::*;

    // AstId: Fn[2A78, 0]
    pub(self) fn fn_in_module;
}

// AstId: Module[C08B, 0]
pub(self) mod outline;

----

Actual:
----
// AstId: Module[709C, 0]
pub(self) mod inline {
    // AstId: Use[0000, 0]
    pub(self) use super::*;

    // AstId: Fn[528A, 0]
    pub(self) fn fn_in_module;
}

// AstId: Module[F422, 0]
pub(self) mod outline;

----

Diff:
----
// AstId: Module[03AE709C, 0]
pub(self) mod inline {
    // AstId: Use[0000, 0]
    pub(self) use super::*;

    // AstId: Fn[2A78528A, 0]
    pub(self) fn fn_in_module;
}

// AstId: Module[C08BF422, 0]
pub(self) mod outline;

----

---- item_tree::tests::modules stdout end ----
---- item_tree::tests::pub_self stdout ----


error: expect test failed
   --> src/tools/rust-analyzer/crates/hir-def/src/item_tree/tests.rs:241:9

Expect:
----
// AstId: Struct[5024, 0]
pub(self) struct S;

----

Actual:
----
// AstId: Struct[DA79, 0]
pub(self) struct S;

----

Diff:
----
// AstId: Struct[5024DA79, 0]
pub(self) struct S;

----

---- item_tree::tests::pub_self stdout end ----
---- item_tree::tests::misc stdout ----


error: expect test failed
   --> src/tools/rust-analyzer/crates/hir-def/src/item_tree/tests.rs:161:9

Expect:
----
// AstId: Static[F7C1, 0]
pub static ST = _;

// AstId: Const[84BB, 0]
pub(self) const _ = _;

#[attr]
#[inner_attr_in_fn]
// AstId: Fn[BE8F, 0]
pub(self) fn f;

// AstId: Trait[9320, 0]
pub(self) trait Tr { ... }

----

Actual:
----
// AstId: Static[A374, 0]
pub static ST = _;

// AstId: Const[D61A, 0]
pub(self) const _ = _;

#[attr]
#[inner_attr_in_fn]
// AstId: Fn[1619, 0]
pub(self) fn f;

// AstId: Trait[1B2D, 0]
pub(self) trait Tr { ... }

----

Diff:
----
// AstId: Static[F7C1A374, 0]
pub static ST = _;

// AstId: Const[84BBD61A, 0]
pub(self) const _ = _;

#[attr]
#[inner_attr_in_fn]
// AstId: Fn[BE8F1619, 0]
pub(self) fn f;

// AstId: Trait[93201B2D, 0]
pub(self) trait Tr { ... }

----

---- item_tree::tests::misc stdout end ----
---- macro_expansion_tests::mbe::concat_spans stdout ----
---
            $tt
        }
    }

    fn main(foo: ()) {
        /* error: unresolved macro unresolved */"helloworld!"#0:Fn[15AE, 0]@236..321#ROOT2024#;
    }
}


----
---
            $tt
        }
    }

    fn main(foo: ()) {
        /* error: unresolved macro unresolved */"helloworld!"#0:Fn[2E90, 0]@236..321#ROOT2024#;
    }
}


----
---
            $tt
        }
    }

    fn main(foo: ()) {
        /* error: unresolved macro unresolved */"helloworld!"#0:Fn[15AE2E90, 0]@236..321#ROOT2024#;
    }
}


----
---
----
#[macro_use]
mod foo;

struct#1:MacroRules[DB0C, 0]@59..65#14336# Foo#0:MacroCall[DB0C, 0]@32..35#ROOT2024#(#1:MacroRules[DB0C, 0]@70..71#14336#u32#0:MacroCall[DB0C, 0]@41..44#ROOT2024#)#1:MacroRules[DB0C, 0]@74..75#14336#;#1:MacroRules[DB0C, 0]@75..76#14336#

----

Actual:
----
#[macro_use]
mod foo;

struct#1:MacroRules[1980, 0]@59..65#14336# Foo#0:MacroCall[1980, 0]@32..35#ROOT2024#(#1:MacroRules[1980, 0]@70..71#14336#u32#0:MacroCall[1980, 0]@41..44#ROOT2024#)#1:MacroRules[1980, 0]@74..75#14336#;#1:MacroRules[1980, 0]@75..76#14336#

----

Diff:
----
#[macro_use]
mod foo;

struct#1:MacroRules[DB0C1980, 0]@59..65#14336# Foo#0:MacroCall[DB0C1980, 0]@32..35#ROOT2024#(#1:MacroRules[DB0C1980, 0]@70..71#14336#u32#0:MacroCall[DB0C1980, 0]@41..44#ROOT2024#)#1:MacroRules[DB0C1980, 0]@74..75#14336#;#1:MacroRules[DB0C1980, 0]@75..76#14336#

----

---- macro_expansion_tests::mbe::token_mapping_across_files stdout end ----
---- macro_expansion_tests::mbe::token_mapping_floats stdout ----


error: expect test failed
   --> src/tools/rust-analyzer/crates/hir-def/src/macro_expansion_tests/mbe.rs:70:9

Expect:
----
// +spans+syntaxctxt
macro_rules! f {
    ($($tt:tt)*) => {
        $($tt)*
    };
}

fn#0:MacroCall[BE8F, 0]@30..32#ROOT2024# main#0:MacroCall[BE8F, 0]@33..37#ROOT2024#(#0:MacroCall[BE8F, 0]@37..38#ROOT2024#)#0:MacroCall[BE8F, 0]@38..39#ROOT2024# {#0:MacroCall[BE8F, 0]@40..41#ROOT2024#
    1#0:MacroCall[BE8F, 0]@50..51#ROOT2024#;#0:MacroCall[BE8F, 0]@51..52#ROOT2024#
    1.0#0:MacroCall[BE8F, 0]@61..64#ROOT2024#;#0:MacroCall[BE8F, 0]@64..65#ROOT2024#
    (#0:MacroCall[BE8F, 0]@74..75#ROOT2024#(#0:MacroCall[BE8F, 0]@75..76#ROOT2024#1#0:MacroCall[BE8F, 0]@76..77#ROOT2024#,#0:MacroCall[BE8F, 0]@77..78#ROOT2024# )#0:MacroCall[BE8F, 0]@78..79#ROOT2024#,#0:MacroCall[BE8F, 0]@79..80#ROOT2024# )#0:MacroCall[BE8F, 0]@80..81#ROOT2024#.#0:MacroCall[BE8F, 0]@81..82#ROOT2024#0#0:MacroCall[BE8F, 0]@82..85#ROOT2024#.#0:MacroCall[BE8F, 0]@82..85#ROOT2024#0#0:MacroCall[BE8F, 0]@82..85#ROOT2024#;#0:MacroCall[BE8F, 0]@85..86#ROOT2024#
    let#0:MacroCall[BE8F, 0]@95..98#ROOT2024# x#0:MacroCall[BE8F, 0]@99..100#ROOT2024# =#0:MacroCall[BE8F, 0]@101..102#ROOT2024# 1#0:MacroCall[BE8F, 0]@103..104#ROOT2024#;#0:MacroCall[BE8F, 0]@104..105#ROOT2024#
}#0:MacroCall[BE8F, 0]@110..111#ROOT2024#



----

Actual:
----
// +spans+syntaxctxt
macro_rules! f {
    ($($tt:tt)*) => {
        $($tt)*
    };
}

fn#0:MacroCall[1619, 0]@30..32#ROOT2024# main#0:MacroCall[1619, 0]@33..37#ROOT2024#(#0:MacroCall[1619, 0]@37..38#ROOT2024#)#0:MacroCall[1619, 0]@38..39#ROOT2024# {#0:MacroCall[1619, 0]@40..41#ROOT2024#
    1#0:MacroCall[1619, 0]@50..51#ROOT2024#;#0:MacroCall[1619, 0]@51..52#ROOT2024#
    1.0#0:MacroCall[1619, 0]@61..64#ROOT2024#;#0:MacroCall[1619, 0]@64..65#ROOT2024#
    (#0:MacroCall[1619, 0]@74..75#ROOT2024#(#0:MacroCall[1619, 0]@75..76#ROOT2024#1#0:MacroCall[1619, 0]@76..77#ROOT2024#,#0:MacroCall[1619, 0]@77..78#ROOT2024# )#0:MacroCall[1619, 0]@78..79#ROOT2024#,#0:MacroCall[1619, 0]@79..80#ROOT2024# )#0:MacroCall[1619, 0]@80..81#ROOT2024#.#0:MacroCall[1619, 0]@81..82#ROOT2024#0#0:MacroCall[1619, 0]@82..85#ROOT2024#.#0:MacroCall[1619, 0]@82..85#ROOT2024#0#0:MacroCall[1619, 0]@82..85#ROOT2024#;#0:MacroCall[1619, 0]@85..86#ROOT2024#
    let#0:MacroCall[1619, 0]@95..98#ROOT2024# x#0:MacroCall[1619, 0]@99..100#ROOT2024# =#0:MacroCall[1619, 0]@101..102#ROOT2024# 1#0:MacroCall[1619, 0]@103..104#ROOT2024#;#0:MacroCall[1619, 0]@104..105#ROOT2024#
}#0:MacroCall[1619, 0]@110..111#ROOT2024#



----

Diff:
----
// +spans+syntaxctxt
macro_rules! f {
    ($($tt:tt)*) => {
        $($tt)*
    };
}

fn#0:MacroCall[BE8F1619, 0]@30..32#ROOT2024# main#0:MacroCall[BE8F1619, 0]@33..37#ROOT2024#(#0:MacroCall[BE8F1619, 0]@37..38#ROOT2024#)#0:MacroCall[BE8F1619, 0]@38..39#ROOT2024# {#0:MacroCall[BE8F1619, 0]@40..41#ROOT2024#
    1#0:MacroCall[BE8F1619, 0]@50..51#ROOT2024#;#0:MacroCall[BE8F1619, 0]@51..52#ROOT2024#
    1.0#0:MacroCall[BE8F1619, 0]@61..64#ROOT2024#;#0:MacroCall[BE8F1619, 0]@64..65#ROOT2024#
    (#0:MacroCall[BE8F1619, 0]@74..75#ROOT2024#(#0:MacroCall[BE8F1619, 0]@75..76#ROOT2024#1#0:MacroCall[BE8F1619, 0]@76..77#ROOT2024#,#0:MacroCall[BE8F1619, 0]@77..78#ROOT2024# )#0:MacroCall[BE8F1619, 0]@78..79#ROOT2024#,#0:MacroCall[BE8F1619, 0]@79..80#ROOT2024# )#0:MacroCall[BE8F1619, 0]@80..81#ROOT2024#.#0:MacroCall[BE8F1619, 0]@81..82#ROOT2024#0#0:MacroCall[BE8F1619, 0]@82..85#ROOT2024#.#0:MacroCall[BE8F1619, 0]@82..85#ROOT2024#0#0:MacroCall[BE8F1619, 0]@82..85#ROOT2024#;#0:MacroCall[BE8F1619, 0]@85..86#ROOT2024#
    let#0:MacroCall[BE8F1619, 0]@95..98#ROOT2024# x#0:MacroCall[BE8F1619, 0]@99..100#ROOT2024# =#0:MacroCall[BE8F1619, 0]@101..102#ROOT2024# 1#0:MacroCall[BE8F1619, 0]@103..104#ROOT2024#;#0:MacroCall[BE8F1619, 0]@104..105#ROOT2024#
}#0:MacroCall[BE8F1619, 0]@110..111#ROOT2024#



----

---
   --> src/tools/rust-analyzer/crates/hir-def/src/macro_expansion_tests/mbe.rs:29:9

Expect:
----
macro_rules! f {
    ( struct $ident:ident ) => {
        struct $ident {
            map: ::std::collections::HashSet<()>,
        }
    };
}

struct#0:MacroRules[BE8F, 0]@58..64#14336# MyTraitMap2#0:MacroCall[BE8F, 0]@31..42#ROOT2024# {#0:MacroRules[BE8F, 0]@72..73#14336#
    map#0:MacroRules[BE8F, 0]@86..89#14336#:#0:MacroRules[BE8F, 0]@89..90#14336# #0:MacroRules[BE8F, 0]@89..90#14336#::#0:MacroRules[BE8F, 0]@91..93#14336#std#0:MacroRules[BE8F, 0]@93..96#14336#::#0:MacroRules[BE8F, 0]@96..98#14336#collections#0:MacroRules[BE8F, 0]@98..109#14336#::#0:MacroRules[BE8F, 0]@109..111#14336#HashSet#0:MacroRules[BE8F, 0]@111..118#14336#<#0:MacroRules[BE8F, 0]@118..119#14336#(#0:MacroRules[BE8F, 0]@119..120#14336#)#0:MacroRules[BE8F, 0]@120..121#14336#>#0:MacroRules[BE8F, 0]@121..122#14336#,#0:MacroRules[BE8F, 0]@122..123#14336#
}#0:MacroRules[BE8F, 0]@132..133#14336#

----

Actual:
----
macro_rules! f {
    ( struct $ident:ident ) => {
        struct $ident {
            map: ::std::collections::HashSet<()>,
        }
    };
}

struct#0:MacroRules[1619, 0]@58..64#14336# MyTraitMap2#0:MacroCall[1619, 0]@31..42#ROOT2024# {#0:MacroRules[1619, 0]@72..73#14336#
    map#0:MacroRules[1619, 0]@86..89#14336#:#0:MacroRules[1619, 0]@89..90#14336# #0:MacroRules[1619, 0]@89..90#14336#::#0:MacroRules[1619, 0]@91..93#14336#std#0:MacroRules[1619, 0]@93..96#14336#::#0:MacroRules[1619, 0]@96..98#14336#collections#0:MacroRules[1619, 0]@98..109#14336#::#0:MacroRules[1619, 0]@109..111#14336#HashSet#0:MacroRules[1619, 0]@111..118#14336#<#0:MacroRules[1619, 0]@118..119#14336#(#0:MacroRules[1619, 0]@119..120#14336#)#0:MacroRules[1619, 0]@120..121#14336#>#0:MacroRules[1619, 0]@121..122#14336#,#0:MacroRules[1619, 0]@122..123#14336#
}#0:MacroRules[1619, 0]@132..133#14336#

----

Diff:
----
macro_rules! f {
    ( struct $ident:ident ) => {
        struct $ident {
            map: ::std::collections::HashSet<()>,
        }
    };
}

struct#0:MacroRules[BE8F1619, 0]@58..64#14336# MyTraitMap2#0:MacroCall[BE8F1619, 0]@31..42#ROOT2024# {#0:MacroRules[BE8F1619, 0]@72..73#14336#
    map#0:MacroRules[BE8F1619, 0]@86..89#14336#:#0:MacroRules[BE8F1619, 0]@89..90#14336# #0:MacroRules[BE8F1619, 0]@89..90#14336#::#0:MacroRules[BE8F1619, 0]@91..93#14336#std#0:MacroRules[BE8F1619, 0]@93..96#14336#::#0:MacroRules[BE8F1619, 0]@96..98#14336#collections#0:MacroRules[BE8F1619, 0]@98..109#14336#::#0:MacroRules[BE8F1619, 0]@109..111#14336#HashSet#0:MacroRules[BE8F1619, 0]@111..118#14336#<#0:MacroRules[BE8F1619, 0]@118..119#14336#(#0:MacroRules[BE8F1619, 0]@119..120#14336#)#0:MacroRules[BE8F1619, 0]@120..121#14336#>#0:MacroRules[BE8F1619, 0]@121..122#14336#,#0:MacroRules[BE8F1619, 0]@122..123#14336#
}#0:MacroRules[BE8F1619, 0]@132..133#14336#

----

---- macro_expansion_tests::mbe::token_mapping_smoke_test stdout end ----
---- macro_expansion_tests::proc_macros::float_attribute_mapping stdout ----


error: expect test failed
   --> src/tools/rust-analyzer/crates/hir-def/src/macro_expansion_tests/proc_macros.rs:259:9

Expect:
----
//+spans+syntaxctxt
#[proc_macros::identity]
fn foo(&self) {
    self.0. 1;
}

fn#0:Fn[8A31, 0]@45..47#ROOT2024# foo#0:Fn[8A31, 0]@48..51#ROOT2024#(#0:Fn[8A31, 0]@51..52#ROOT2024#&#0:Fn[8A31, 0]@52..53#ROOT2024#self#0:Fn[8A31, 0]@53..57#ROOT2024# )#0:Fn[8A31, 0]@57..58#ROOT2024# {#0:Fn[8A31, 0]@59..60#ROOT2024#
    self#0:Fn[8A31, 0]@65..69#ROOT2024# .#0:Fn[8A31, 0]@69..70#ROOT2024#0#0:Fn[8A31, 0]@70..71#ROOT2024#.#0:Fn[8A31, 0]@71..72#ROOT2024#1#0:Fn[8A31, 0]@73..74#ROOT2024#;#0:Fn[8A31, 0]@74..75#ROOT2024#
}#0:Fn[8A31, 0]@76..77#ROOT2024#
----

Actual:
----
//+spans+syntaxctxt
#[proc_macros::identity]
fn foo(&self) {
    self.0. 1;
}

fn#0:Fn[2F37, 0]@45..47#ROOT2024# foo#0:Fn[2F37, 0]@48..51#ROOT2024#(#0:Fn[2F37, 0]@51..52#ROOT2024#&#0:Fn[2F37, 0]@52..53#ROOT2024#self#0:Fn[2F37, 0]@53..57#ROOT2024# )#0:Fn[2F37, 0]@57..58#ROOT2024# {#0:Fn[2F37, 0]@59..60#ROOT2024#
    self#0:Fn[2F37, 0]@65..69#ROOT2024# .#0:Fn[2F37, 0]@69..70#ROOT2024#0#0:Fn[2F37, 0]@70..71#ROOT2024#.#0:Fn[2F37, 0]@71..72#ROOT2024#1#0:Fn[2F37, 0]@73..74#ROOT2024#;#0:Fn[2F37, 0]@74..75#ROOT2024#
}#0:Fn[2F37, 0]@76..77#ROOT2024#
----

Diff:
----
//+spans+syntaxctxt
#[proc_macros::identity]
fn foo(&self) {
    self.0. 1;
}

fn#0:Fn[8A312F37, 0]@45..47#ROOT2024# foo#0:Fn[8A312F37, 0]@48..51#ROOT2024#(#0:Fn[8A312F37, 0]@51..52#ROOT2024#&#0:Fn[8A312F37, 0]@52..53#ROOT2024#self#0:Fn[8A312F37, 0]@53..57#ROOT2024# )#0:Fn[8A312F37, 0]@57..58#ROOT2024# {#0:Fn[8A312F37, 0]@59..60#ROOT2024#
    self#0:Fn[8A312F37, 0]@65..69#ROOT2024# .#0:Fn[8A312F37, 0]@69..70#ROOT2024#0#0:Fn[8A312F37, 0]@70..71#ROOT2024#.#0:Fn[8A312F37, 0]@71..72#ROOT2024#1#0:Fn[8A312F37, 0]@73..74#ROOT2024#;#0:Fn[8A312F37, 0]@74..75#ROOT2024#
}#0:Fn[8A312F37, 0]@76..77#ROOT2024#
----

---- macro_expansion_tests::proc_macros::float_attribute_mapping stdout end ----

failures:
---
    macro_expansion_tests::proc_macros::float_attribute_mapping

test result: FAILED. 453 passed; 12 failed; 1 ignored; 0 measured; 0 filtered out; finished in 295.35ms

error: test failed, to rerun pass `-p hir-def --lib`
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 1:55:10
  local time: Fri Dec  5 11:44:42 UTC 2025
  network time: Fri, 05 Dec 2025 11:44:42 GMT
##[error]Process completed with exit code 1.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/19958935918/job/57244255837)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/19958935918/job/57244255837","builder_name":"checks-actions"} -->

jieyouxu commented:
Yeah, if 32-bit builds are not intended to be shipped / tested, we can skip these tests for 32-bit platforms (or rather, run them only on 64-bit platforms).

jieyouxu committed:
bootstrap: skip r-a main tests against 32-bit targets

We don't really have a good way to get the target pointer width in
bootstrap (yet), so I had to resort to hacky target tuple substring
matches.

This commit currently only gates on `i686-*` targets, which are 32-bits,
and does not try to gate on `i586-*` targets (I need to run try jobs to
check if r-a tests actually get run.)

jieyouxu commented:
Random CI Job :tm: go
@bors try jobs=dist-i586-gnu-i586-i686-musl
